crates/kubelet/src/pod/queue.rs

```diff
@@ -19,79 +30,166 @@ use crate::provider::Provider;
+    handlers: HashMap<String, tokio::sync::mpsc::Sender<WatchEvent<KubePod>>>,
+    client: KubeClient,
 }
 
-struct Worker {
-    sender: watch::Sender<WatchEvent<KubePod>>,
-    _worker: JoinHandle<()>,
-}
+impl<P: 'static + Provider + Sync + Send> Queue<P> {
+    pub fn new(provider: Arc<P>, client: KubeClient) -> Self {
+        Queue {
+            provider,
+            handlers: HashMap::new(),
+            client,
+        }
+    }
 
-impl Worker {
-    fn create<P>(
+    async fn run_pod(
+        &self,
         initial_event: WatchEvent<KubePod>,
-        provider: Arc<P>,
-        mut error_sender: Sender<(KubePod, anyhow::Error)>,
-    ) -> Self
-    where
-        P: 'static + Provider + Sync + Send,
-    {
-        let (sender, mut receiver) = watch::channel(initial_event);
-        let worker = tokio::spawn(async move {
+    ) -> anyhow::Result<tokio::sync::mpsc::Sender<WatchEvent<KubePod>>> {
+        let (sender, mut receiver) = tokio::sync::mpsc::channel::<WatchEvent<KubePod>>(16);
+
+        let pod_deleted = Arc::new(RwLock::new(false));
+        let check_pod_deleted = Arc::clone(&pod_deleted);
+
+        match initial_event {
+            WatchEvent::Added(pod) => {
+                let task_client = self.client.clone();
+                let pod = Pod::new(pod);
+
+                let mut pod_state: P::PodState = self.provider.initialize_pod_state(&pod).await?;
+                tokio::spawn(async move {
+                    let state: P::InitialState = Default::default();
+                    let name = pod.name().to_string();
+
+                    let check = async {
+                        loop {
+                            tokio::time::delay_for(std::time::Duration::from_secs(1)).await;
+                            if *check_pod_deleted.read().await {
+                                break;
+                            }
+                        }
+                    };
+
+                    tokio::select! {
+                        result = run_to_completion(&task_client, state, &mut pod_state, &pod) => match result {
+                            Ok(()) => info!("Pod {} state machine exited without error", name),
+                            Err(e) => {
+                                error!("Pod {} state machine exited with error: {:?}", name, e);
+                                let api: kube::Api<KubePod> = kube::Api::namespaced(task_client.clone(), pod.namespace());
+                                let patch = serde_json::json!(
+                                    {
+                                        "metadata": {
+                                            "resourceVersion": "",
+                                        },
+                                        "status": {
+                                            "phase": Phase::Failed,
+                                            "reason": format!("{:?}", e),
+                                            "containerStatuses": Vec::<()>::new(),
```

Comment:
These are still a TODO, right? Just making sure we don't lose track

Reply:
Yeah, I figured this was something we can iterate on.